[stable/spinnaker] Leverage Halyard for installation

viglesiasce opened the pull request:
This pull requests removes most of the installation logic from the chart and instead leverages [Halyard](https://www.spinnaker.io/reference/halyard/), the Spinnaker project's tool for creating, configuring, and managing Spinnaker deployments.

Benefits:
1. Much easier customization
2. Less burden to update Spinnaker version (one liner)
3. Spinnaker Kubernetes v2 provider support
4. Decouples Jenkins installation
5. 2000 less lines of code to maintain

Commit:
Use Halyard for Spinnaker installation

Commit:
Bump spinnaker version in Chart.yaml

Commit:
Add whitespace in helpers

Commit:
Remove dependency on Jenkins

Commit:
Fix cleanup

Commit:
Fix NOTES command for exec into hal

Commit:
No longer need override_config_map helper

Commit:
Simplify cleanup job

Commit:
Remove halyard-pvc empty file

Commit:
Update README

lwander commented on stable/spinnaker/templates/statefulsets/halyard.yaml:
```diff
@@ -0,0 +1,70 @@
+        command:
+        - bash
+        - -c
+        - "cp /opt/halyard/config/* /tmp/config && printf 'server:\n  address: 0.0.0.0\n' > /tmp/config/halyard-local.yml"
```
`server.address: 0.0.0.0\n` might be a little cleaner

viglesiasce replied:
Done! Good call.

lwander reviewed (approved):
Nice!

Commit:
Add lwander as maintainer of Spinnaker chart

Commit:
Make halyard-local config simpler

Commit:
Remove type: Opaque from kubeconfig map

Commit:
Mount GCS key on Halyard statefulset

Commit:
Fix credentials when using kubeConfig.enabled

eaceaser reviewed (commented):
https://github.com/kubernetes/charts/pull/6407/files#diff-13a5d824f78f72778667d929a7c83c4dR64
https://github.com/kubernetes/charts/pull/6407/files#diff-61d64636826856804e1519cd255d54e8R2

Unless there are some changes in store to have this chart also deploy redis for caching at some point, I think we can remove the redis values and requirement as part of this too!

https://github.com/kubernetes/charts/pull/6407/files#diff-91f0ff1f45e99083851f8f4f1b909dbbR9

And the README copy would be outdated too I think

eaceaser commented on stable/spinnaker/values.yaml:
```diff
@@ -71,25 +51,9 @@ deck:
-  allowedOriginsPattern: '^https?://(?:localhost|127.0.0.1|[^/]+\.example\.com)(?::[1-9]\d*)?/?$'
-
 # Bucket to use when storing config data in S3 compatible storage
 storageBucket: spinnaker
```
Would it be clearer to denormalize (for lack of a better term) `storageBucket` into the various storage providers' config blocks? Unless it gets used somewhere that I'm not seeing it seems like it should be configured alongside the storage provider, rather than outside of it as a top level key.

Commit:
Allow custom config to be provided at install time.

Commit:
Fix port-forward commands in README

Commit:
Configure Halyard's RBAC

Commit:
Use redis deployed by helm

Commit:
Use Deck to route /gate requests

Commit:
Fix RBAC in default configuration

eaceaser committed:
Scope template helper names

eaceaser committed:
Helper for resource metadata

dpujadas committed:
Enable spinnakerFeatureFlags

dpujadas committed:
Fix deck service name

Commit:
Merge pull request #11 from eaceaser/spin-v1.0.0-updates

Scope helper names, add resource label helpers

Commit:
Merge branch 'spin-v1.0.0' into spin-v1.0.0-fix-ingress

Commit:
Merge pull request #13 from GekoCloud/spin-v1.0.0-fix-ingress

Fix deck service name

eaceaser committed:
Basic S3 support.

setting access keys is optional, if they are unset, the implicit
credentials from your environment will be used.

eaceaser committed:
Smallish updates.

* Runs the install script with bash -xe for log output.
* Makes the install script somewhat idempotent in the case of chart
  upgrades on an existing halyard config state.
* Fully qualifies serviceaccount names.
* Whitespace fixes.

costimuraru commented on stable/spinnaker/README.md:
```diff
@@ -55,5 +55,38 @@ for Spinnaker. If you want to add arbitrary clusters need to do the following:
+If you have known set of commands that you'd like to run after the base config steps or if
+you'd like to override some settings before the Spinnaker deployment is applied, you can enable
+the `halyard.additionalConfig.enabled` flag. You will need to create a config map that contains a key
+contianing the `hal` commands you'd like to run. You can set the key via the config map name via `halyard.additionalConfig.configMapName` and the key via `halyard.additionalConfig.configMapKey`. The `DAEMON_ENDPOINT` environment variable can be used in your custom commands to
```
Typo; `contianing` -> `containing`

costimuraru commented on stable/spinnaker/README.md:
````diff
@@ -55,5 +55,38 @@ for Spinnaker. If you want to add arbitrary clusters need to do the following:
+contianing the `hal` commands you'd like to run. You can set the key via the config map name via `halyard.additionalConfig.configMapName` and the key via `halyard.additionalConfig.configMapKey`. The `DAEMON_ENDPOINT` environment variable can be used in your custom commands to
+get a prepopulated URL that points to your Halyard daemon within the cluster. For example:
+
+```shell
````
Should this shell command be present in a ConfigMap map that we manually create?
If so, could we also provide a sample `yaml` for this ConfigMap, to make things a little bit easier to understand?

I'm imagining that we would need to do something like:

`vi my-custom-hal-cmds-configmap.yaml`

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: my-custom-halyard-config
data:
  custom_commands.sh: |
    hal --daemon-endpoint $DAEMON_ENDPOINT config security authn oauth2 enable
```

`kubectl apply -f my-custom-hal-cmds-configmap.yaml`

And then run 
```bash
helm install stable/spinnaker 
    --set halyard.additionalConfig.configMapName=my-custom-halyard-config 
    --set halyard.additionalConfig.configMapKey=custom_commands.sh
```

Is that right?

Commit:
Merge pull request #15 from eaceaser/spin-v1.0.0-updates-2

Spinnaker chart smallish updates

Commit:
Merge branch 'spin-v1.0.0' into spin-v1.0.0-s3

Commit:
Merge pull request #14 from eaceaser/spin-v1.0.0-s3

Basic S3 support.

dpujadas committed:
Enable 'jobs' feature flag by default

unguiculus commented on stable/spinnaker/templates/statefulsets/halyard.yaml:
```diff
@@ -0,0 +1,110 @@
+  selector:
+    app: {{ template "spinnaker.fullname" . }}
+    component: halyard
+---
```
One file per resource, please.

unguiculus commented on stable/spinnaker/templates/statefulsets/halyard.yaml:
```diff
@@ -0,0 +1,110 @@
+{{ include "spinnaker.standard-labels" . | indent 8 }}
+        component: halyard
+    spec:
+      serviceAccountName: {{ template "spinnaker.fullname" . }}-halyard
```
Could you update the chart to use our RBAC best practices?
https://github.com/kubernetes/helm/blob/master/docs/chart_best_practices/rbac.md

viglesiasce replied:
Updated. PTAL

unguiculus commented on stable/spinnaker/templates/statefulsets/halyard.yaml:
```diff
@@ -0,0 +1,110 @@
+          mountPath: /home/spinnaker
+        - name: halyard-config
+          mountPath: /opt/halyard/config
+  volumeClaimTemplates:
```
Please apply our persistence best practices:
https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md#persistence

Here's an example for a statefulset:
https://github.com/kubernetes/charts/blob/master/stable/mongodb-replicaset/templates/mongodb-statefulset.yaml#L313-L339

unguiculus reviewed (changes requested):
I made a few comments regarding best practices. Since this is a major refactoring, you may want to apply these to the existing stuff as well where it makes sense.

Commit:
Merge branch 'spin-v1.0.0' into spin-v1.0.0-feature-flags

Commit:
Merge pull request #12 from GekoCloud/spin-v1.0.0-feature-flags

Optionally enable Spinnaker feature flags

Commit:
Fix typo in README

paulczar commented on stable/spinnaker/values.yaml:
```diff
@@ -109,27 +85,20 @@ minio:
-#   https://www.spinnaker.io/setup/ci/jenkins/
-jenkins:
-  enabled: true
-  Master:
```
does this need to be capitalized for an external dependency?  otherwise it would be nice to lowercase it to remain consistent.

viglesiasce replied:
This is indeed how the Jenkins charts naming works. That said, the dependency on Jenkins has been removed.

Commit:
Follow RBAC best practices for chart

Commit:
Remove trailing space

paulczar committed:
update helm dependencies

Update minio and redis deps to be their current releases

paulczar committed:
support username/password for docker registry

Commit:
Merge pull request #17 from paulczar/update-minio-redis-deps

update helm dependencies

paulczar committed:
fix variable condition for registrysecret

paulczar committed:
remove rogue sleep command from debugging

paulczar committed:
add readme info about using secret for registry passwords

Commit:
Merge pull request #18 from paulczar/repo-creds

Allow user to specify registry credentials in values or via a premade secret.

paulczar reviewed (approved):
I've tested this thoroughly and it works well, a big step up from the previous non-halyard charts.

eaceaser committed:
fix writing the secret key to hal's config

Commit:
Merge pull request #19 from eaceaser/spin-v1.0.0-s3fix

Fix for S3 credentials config

Commit:
Enable artifacts by default

Commit:
Allow Halyard and Spinnaker SAs to be configurable

Commit:
Use custom SA in halyard SS if passed

Commit:
Default to version 1.8.4 of Spinnaker

Commit:
Simplify ingress values, fix nodeports for svcs

sc250024 committed:
Slight change; moving 'host' to be under Ingress resource

Commit:
Merge pull request #21 from sc250024/spin-v1.0.0

Moving 'hosts' to ingress

Commit:
Merge branch 'master' into spin-v1.0.0

Commit:
Updater Spinnaker to 1.8.5

scottrigby commented on stable/spinnaker/templates/rbac/halyard-sa.yaml:
```diff
@@ -0,0 +1,13 @@
+  namespace: {{ .Release.Namespace }}
+  labels:
+{{ include "spinnaker.standard-labels" . | indent 4 }}
+{{- end }}
```
Needs newline at EOF

scottrigby replied:
👍

scottrigby commented on stable/spinnaker/templates/rbac/spinnaker-sa.yaml:
```diff
@@ -0,0 +1,26 @@
+  # service account.
+  name: default
+{{- end }}
+{{- end }}
```
Needs newline at EOF

scottrigby replied:
👍

scottrigby commented on stable/spinnaker/templates/secrets/registry.yaml:
```diff
@@ -1,12 +1,14 @@
+  {{- range $index, $account := .Values.dockerRegistries }}
+  {{ $account.name }}: {{ default "" $account.password | b64enc | quote }}
+  {{- end }}
+{{- end }}
```
Needs newline at EOF

scottrigby replied:
👍

eaceaser committed:
Make nodeport exposure idempotent

scottrigby commented on stable/spinnaker/templates/services/halyard.yaml:
```diff
@@ -0,0 +1,15 @@
+  clusterIP: None
+  selector:
+    app: {{ template "spinnaker.fullname" . }}
+    component: halyard
```
Needs newline at EOF

scottrigby replied:
👍

scottrigby reviewed (commented):
Small nits

Commit:
Add newline

Commit:
Add newline

Commit:
Add newline

Commit:
Add newline

Commit:
Merge pull request #24 from eaceaser/spin-v1.0.0-nodeport-fix

Make nodeport exposure idempotent

scottrigby reviewed (approved):
Re https://github.com/helm/charts/pull/6407#issuecomment-414082524 and https://github.com/helm/charts/pull/6407#issuecomment-415497659, it seems to me the RBAC follows the best practices guidelines. `rbac.create` is configurable, as is `serviceAccount.create`. The service account names (`serviceAccount.halyardName` and `serviceAccount.spinnakerName` are also configurable, and a comment indicates that if left blank it's auto-generated from the release full name). The Spinnaker service account hard-codes `default` with this note:
>Clouddriver does not currently allow config of its service account.

Personally I would say this looks good to merge. The MAJOR version is bumped, and it's fully up to date (a moment ago recently merged in updates from `master`).

@viglesiasce One request though, if there's any suggestions you have to help make https://github.com/helm/helm/blob/master/docs/chart_best_practices/rbac.md more clear, that would help #3011. Anyway thanks this looks 💯 to me
